lightning-invoice/src/payment.rs

```diff
@@ -399,20 +465,21 @@ where
 	fn retry_payment(
 		&self, payment_id: PaymentId, payment_hash: PaymentHash, params: &RouteParameters
 	) -> Result<(), ()> {
-		let max_payment_attempts = self.retry_attempts.0 + 1;
-		let attempts = *self.payment_cache.lock().unwrap()
-			.entry(payment_hash)
-			.and_modify(|attempts| *attempts += 1)
-			.or_insert(1);
-
-		if attempts >= max_payment_attempts {
-			log_trace!(self.logger, "Payment {} exceeded maximum attempts; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+		let mut payment_cache = self.payment_cache.lock().unwrap();
+		let mut attempts = payment_cache.entry(payment_hash).or_insert(PaymentAttempts::new());
+
+		if !self.retry.is_retryable_now(&attempts) {
+			log_trace!(self.logger, "Payment {} exceeded maximum attempts; not retrying ({:?})", log_bytes!(payment_hash.0), attempts);
 			return Err(());
 		}
 
+		attempts.retry_count += 1;
+		// we have to manually drop payment_cache to unlock its mutex to avoid deadlock with scorer.lock().
+		let attempts = attempts.clone();
+		core::mem::drop(payment_cache);
 		#[cfg(feature = "std")] {
 			if has_expired(params) {
-				log_trace!(self.logger, "Invoice expired for payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+				log_trace!(self.logger, "Invoice expired for payment {}; not retrying ({:?})", log_bytes!(payment_hash.0), attempts);
```

Comment:
Hmm... The attempts won't be accurate here since you've incremented earlier (and in `Debug`) but no retry occurs here. Similarly below.

Would it simplify things if we made `PaymentAttempts::retry_count` simply a `count` representing the number of attempts up until the retry? I believe that is what was done previously and as is still documented on the hash map:

```rust
/// Caches the overall attempts at making a payment, which is updated prior to retrying.
```

Essentially, the attempt is counted only once we get an event back on whether a path was successful or not. So renaming to `count`, unconditionally incrementing in this function, and not incrementing in `Debug` should give the correct number. Would need to modify `is_retryable_now`, too, IIUC.

Then you don't need to worry about the lock since you can increment in the above look-up, clone, and drop the lock implicitly in one statement similarly to the code prior to the change.

Reply:
ah u are right that the log in this branch(`has_expired` true) is not accurate. But can we simply move this has_expire check before `is_retryable_now?` I ran test in local after this change and all test case passed as well.



As per your suggestion, I believed that was my initial approach(got forced-pushed) but I decided to change it to only store retry_count because this part( https://github.com/lightningdevkit/rust-lightning/pull/1418#discussion_r848721157) is confusing to others.

If we want to avoid the confusing part in my previous forced-pushed commit, does it mean, does it mean we need to change the logic in `Retry::is_retryable_now` to be aware of the fact count always get incremented by 1 before calling retry_payment. I found that it's actually more confusing as now `Retry::is_retryable_now` is not a context-free function especially now `is_retryable_now` is also reused inside `PaymentSendFailure::AllFailedRetrySafe`

Reply:
Hmm... I think it should work as expected. For `AllFailedRetrySafe` the number of attempts before increment (i.e., `retry_count`) and max retries are both 0, so an error is returned here:

https://github.com/lightningdevkit/rust-lightning/blob/9fbafd4b6cc7a290fbd952d8baf58d6017fc3078/lightning-invoice/src/payment.rs#L370-L372

And for `PartialFailure` we hit the `retry_payment` case where the number of attempts is incremented and compared against maximum number of attempts (i.e. max retries plus one).

https://github.com/lightningdevkit/rust-lightning/blob/9fbafd4b6cc7a290fbd952d8baf58d6017fc3078/lightning-invoice/src/payment.rs#L402-L411

Reply:
@jkczyz Thank you so much for breaking it down. I will break the last commit into 2 as per your suggestion so it's easier for us to see if the change make sense or not(And I will also look into if the original approach result in some incorrect log msg) . However, I will probably only have time this weekend🙏🏽

Reply:
@jkczyz I have made changes to restore the previous behavior and store attempts instead of retry attempts. After spending more time understanding the code, I actually think it's cleaner to keep in the original way :)